File: bin/new-problem.rb
#!/usr/bin/env ruby

def die(message)
  puts "error: #{message}"
  exit 1
end

die 'missing number argument' unless ARGV.count > 0
die "#{ARGV[0]} is not a number" unless ARGV[0] =~ /^[0-9]+$/

number = "%03d" % ARGV[0].to_i
filename = "Problem#{number}"
modulename = "ProjectEuler.#{filename}"
filepath = "src/ProjectEuler/#{filename}.hs"
testpath = "test/ProjectEuler/#{filename}Spec.hs"
cabalpath = 'project-euler.cabal'

die "#{filepath} exists" if File.exist? filepath
die "#{testpath} exists" if File.exist? testpath

file_template=%(module #{modulename} (solve) where

solve :: Integer
solve = #{number}
)

test_template=%(module #{modulename}Spec (main, spec) where

import Test.Hspec
import #{modulename}

main :: IO ()
main = hspec spec

spec :: Spec
spec = parallel $
  describe "solve" $
    it "returns #{number}" $
      solve `shouldBe` #{number}
)

File.open(filepath, 'w') { |f| f.write(file_template) }
File.open(testpath, 'w') { |f| f.write(test_template) }

lines = File.foreach(cabalpath).to_a
lines.reverse.each.with_index do |line, i|
  if line =~ /ProjectEuler.Problem/
    lines.insert(
      lines.length - i,
      "                     , ProjectEuler.Problem#{number}\n"
    )
    break
  end
end

File.open(cabalpath, 'w') { |f| f.write(lines.join('')) }

File: scripts/NewProblem.hs
module Main (main) where

import Control.Monad
import Data.List
import Data.Maybe
import System.Directory
import System.Environment
import System.Exit
import Text.Printf
import Text.Read

(%) :: PrintfArg a => String -> a -> String
(%) f a = printf f a :: String

fileTemplate :: String -> String -> String
fileTemplate modulename number = unlines 
  [ "module %s (solve) where" % modulename
  , ""
  , "solve :: Integer"
  , "solve = %s" % number
  ]

testTemplate :: String -> String -> String
testTemplate modulename number = unlines
  [ "module %sSpec (main, spec) where" % modulename
  , ""
  , "import Test.Hspec"
  , "import %s" % modulename
  , ""
  , "main :: IO ()"
  , "main = hspec spec"
  , ""
  , "spec :: Spec"
  , "spec = parallel $"
  , "  describe \"solve\" $"
  , "    it \"returns %s\" $" % number
  , "      solve `shouldBe` %s" % number
  ]

insertAtIndex :: Int -> a -> [a] -> [a]
insertAtIndex i x xs = join [ys, [x], zs]
  where
    (ys, zs) = splitAt i xs

main :: IO ()
main = do
  args <- getArgs

  when (null args) (die "Number argument missing.")

  let a = head args
      m = readMaybe a :: Maybe Int

  when (isNothing m) (die $ "%s is not a number." % a)

  let number     = "%03d"                        % fromJust m
      filename   = "Problem%s"                   % number
      modulename = "ProjectEuler.%s"             % filename
      cabalpath  = "project-euler.cabal"
      filepath   = "src/ProjectEuler/%s.hs"      % filename
      testpath   = "test/ProjectEuler/%sSpec.hs" % filename
      files      = [(filepath, fileTemplate), (testpath, testTemplate)]

  forM_ files (\(path, template) -> do
    writeFile path (template modulename number)
    printf "Created %s.\n" path)

  printf $ "Add %s to library.exposed-modules.\n" % modulename
  printf $ "Add %sSpec to test-suite.other-modules.\n" % modulename
